sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs

```diff
@@ -0,0 +1,187 @@
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+            protected override bool IsTrackedPropertyName(string propertyName) => "XmlResolver".Equals(propertyName);
+
+            internal override ImmutableArray<KnownType> TrackedTypes { get; } = ImmutableArray.Create(
+                KnownType.System_Xml_XmlDocument,
+                KnownType.System_Xml_XmlDataDocument,
+                KnownType.System_Configuration_ConfigXmlDocument,
+                KnownType.Microsoft_Web_XmlTransform_XmlFileInfoDocument,
+                KnownType.Microsoft_Web_XmlTransform_XmlTransformableDocument
+            );
+
+            protected override bool CtorInitializesTrackedPropertyWithAllowedValue(ArgumentListSyntax argumentList, SemanticModel semanticModel) =>
+                NetFrameworkVersion switch
+                {
+                    NetFrameworkVersion.After45 => true,
+                    NetFrameworkVersion.Between4And45 => false,
+                    NetFrameworkVersion.Below4 => false,
+                    _ => true
+                };
+
+            // we do not expect any constant values for XmlResolver
+            protected override bool IsAllowedValue(object constantValue) => true;
+
+            protected override bool IsAllowedValue(ISymbol symbol) =>
+                !IsUnsafeXmlResolverConstructor(symbol) &&
+                !symbol.GetSymbolType().IsAny(UnsafeXmlResolvers);
+
+            protected override void CompilationAction(Compilation compilation)
+            {
+                NetFrameworkVersion = VersionProvider.GetDotNetFrameworkVersion(compilation);
+            }
+
+            internal XmlDocumentShouldBeSafe(INetFrameworkVersionProvider netFrameworkVersionProvider)
+            {
+                VersionProvider = netFrameworkVersionProvider;
+            }
+
+            private bool IsUnsafeXmlResolverConstructor(ISymbol symbol) =>
+                symbol.Kind == SymbolKind.Method &&
+                symbol.ContainingType.GetSymbolType().IsAny(UnsafeXmlResolvers);
+        }
+
+        [SuppressMessage("AnalyzerCorrectness", "RS1001:Missing diagnostic analyzer attribute.", Justification = "This is not a different diagnostic.")]
+        private class XmlTextReaderShouldBeSafe : ObjectShouldBeInitializedCorrectlyBase
+        {
+            private NetFrameworkVersion NetFrameworkVersion;
+
+            private INetFrameworkVersionProvider VersionProvider;
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+            private static readonly ISet<string> TrackedProperties = new HashSet<string>
+            {
+                "XmlResolver", // should be null
+                "DtdProcessing", // should not be Parse
+                "ProhibitDtd" // should be true in .NET 3.5
+            }.ToImmutableHashSet();
```

Comment:
In order to avoid creation of an intermediate HashSet we can use `ImmutableHashSet.Create("XmlResolver", "DtdProcessing", "ProhibitDtd");`